src/java/org/apache/cassandra/index/sai/disk/v1/sortedterms/SortedTermsWriter.java

```diff
@@ -0,0 +1,216 @@
+
+/**
+ * Writes an ordered sequence of terms for use with {@link SortedTermsReader}.
+ * <p>
+ * Terms must be added in lexicographical ascending order.
+ * Terms can be of varying lengths.
+ *
+ * <p>
+ * Important implementation note: SAI blocked packed readers are slow,
+ * and Lucene MonotonicBlockPackedReader is slow.  Using them
+ * will cause this class to slow considerably.
+ *
+ * For documentation of the underlying on-disk data structures, see the package documentation.
+ *
+ * @see SortedTermsReader
+ * @see org.apache.cassandra.index.sai.disk.v1.sortedterms
+ */
+@NotThreadSafe
+public class SortedTermsWriter implements Closeable
+{
+    // The TERMS_DICT_ constants allow for quickly determining the id of the current block based on a point id
+    // or to check if we are exactly at the beginning of the block.
+    // Terms data are organized in blocks of (2 ^ TERMS_DICT_BLOCK_SHIFT) terms.
+    // The blocks should not be too small because they allow prefix compression of
+    // the terms except the first term in a block.
+    // The blocks should not be too large because we can't just ranfomly jump to the term inside the block,
+    // but we have to iterate through all the terms from the start of the block.
+    static final int TERMS_DICT_BLOCK_SHIFT = 4;
+    static final int TERMS_DICT_BLOCK_SIZE = 1 << TERMS_DICT_BLOCK_SHIFT;
+    static final int TERMS_DICT_BLOCK_MASK = TERMS_DICT_BLOCK_SIZE - 1;
+
+    private final IncrementalDeepTrieWriterPageAware<Long> trieWriter;
+    private final IndexOutput trieOutput;
+    private final IndexOutput termsOutput;
+    private final NumericValuesWriter offsetsWriter;
+    private final String componentName;
+    private final MetadataWriter metadataWriter;
+
+    private BytesRefBuilder prevTerm = new BytesRefBuilder();
+    private BytesRefBuilder tempTerm = new BytesRefBuilder();
+
+    private final long bytesStartFP;
+
+    private int maxLength = -1;
+    private long pointId = 0;
+
+    /**
+     * Creates a new writer.
+     * <p>
+     * It does not own the components, so you must close the components by yourself
+     * after you're done with the writer.
+     *
+     * @param componentName the component name for the SortedTermsMeta
+     * @param metadataWriter the MetadataWriter for storing the SortedTermsMeta
+     * @param termsData where to write the prefix-compressed terms data
+     * @param termsDataBlockOffsets  where to write the offsets of each block of terms data
+     * @param trieWriter where to write the trie that maps the terms to point ids
+     */
+    public SortedTermsWriter(String componentName,
+                             MetadataWriter metadataWriter,
+                             IndexOutput termsData,
+                             NumericValuesWriter termsDataBlockOffsets,
+                             IndexOutputWriter trieWriter) throws IOException
```

Comment:
```suggestion
                             IndexOutputWriter trieOutput) throws IOException
```
Seems less confusing, w/ other things also named `trieWriter`?